src/http-gateways/trustless-gateway.md

```diff
@@ -8,28 +8,36 @@ editors:
 
 # Trustless Gateway Specification
 
-Trustless Gateway is a minimal _subset_ of :cite[path-gateway]
+Trustless Gateway is a _subset_ of :cite[path-gateway]
 that allows light IPFS clients to retrieve data behind a CID and verify its
 integrity without delegating any trust to the gateway itself.
 
 The minimal implementation means:
 
-- data is requested by CID, only supported path is `/ipfs/{cid}`
-- no path traversal or recursive resolution, no UnixFS/IPLD decoding server-side
 - response type is always fully verifiable: client can decide between a raw block or a CAR stream
+- no UnixFS/IPLD decoding server-side
+- for CAR files:
+  - the behavior is identical to :cite[path-gateway]
+- for raw blocks:
+  - data is requested by CID, only supported path is `/ipfs/{cid}`
+  - no path traversal or recursive resolution
 
 # HTTP API
 
 A subset of "HTTP API" of :cite[path-gateway].
 
-## `GET /ipfs/{cid}[?{params}]`
+## `GET /ipfs/{cid}[/{path}][?{params}]`
 
-Downloads data at specified CID.
+Downloads data at specified **immutable** content path.
 
-## `HEAD /ipfs/{cid}[?{params}]`
+Path parameters are only permitted for requests that specify CAR format. For RAW requests, only `GET /ipfs/{cid}[?{params}]` is supported
+
+## `HEAD /ipfs/{cid}[/{path}][?{params}]`
 
 Same as GET, but does not return any payload.
 
+Path parameters are only permitted for requests that specify CAR format. For RAW requests, only `HEAD /ipfs/{cid}[?{params}]` is supported
+
```

Comment:
## Proposal for feature detection via HTTP OPTIONS

Old gateways will ignore graph parameters and return CAR with the entire DAG.
We need a light mechanism for clients to detect which gateway supports partial CARs.

There is a good prior art for this in IPNI, where `OPTIONS` method returns headers when a specific feature is supported, and the lack of header means no support. 

This way light clients can adjust retrieval strategy and if graph is not available, do brute-force retrieval like the one in RAPIDE.


```suggestion

## `OPTIONS /ipfs/[{cid}[/{path}]]`

A client may discover the list of additional features supported by the trustless gateway server via sending `OPTIONS` request.

In response, the server may include HTTP headers:

- `Ipfs-Gateway-Format-Car-Verifiable-Paths` set to `true` indicates `/ipfs/cid/dir/file.jpg?format=car` will return a CAR with minimal set of blocks required for verifying the entire content path.
- `Ipfs-Gateway-Format-Car-Scopes` header key, with value as the comma separated list of `car-scope` parameters supported. 
  - For example, `Ipfs-Gateway-Format-Car-Scopes: block, all` means `/ipfs/cid/dir/file.jpg?format=car&car-scope=block` can be used to return all blocks for parent directories and only the root block for the `file.jpg`. 
- `Ipfs-Gateway-Format-Car-Bytes` set to `true` to indicate range requests via `?format=car&bytes=from:to` are supported.


The absence of a header implies that the server does not offer support  specific feature and that it will always return the entire DAG behind the requested CID.
```

Reply:
Would appreciate feedback on ergonomics of doing feature-detection this way from  potential client/server implementers  (cc @hannahhoward @guanzo @olizilla  @hacdias @John-LittleBearLabs @fabricedesre   for visibility)

Reply:
> This request if often sent by web browser anyway, so we would not be introducing much overhead.

The OPTIONS request/response isn't accessible to userland JS, so browser clients can't see these resposne headers and know what features are supported.

Reply:
I think this seems like a fantastic idea.

I don't know if HTTP headers are easier/better than doing something like JSON in the body, but it would certainly have worked for me up to this point. I suppose JSON (or similar) would have the option to arrange the data differently, e.g. `"Formats": [ "car", "raw", "dag-cbor", "dag-json" ]`. I guess the downside is that right now I think it's theoretically possible to have a useful client that doesn't actually know how to parse JSON, and headers might be a more primitive thing to read.

Would we want to use a feature-specific prefix? Like `Ipfs-Gateway-Feature-` so a client could _note_ the presence of features it's unaware of (distinguishing it from headers that aren't feature announcements)? I'm not sure exactly what one could do with that other than log it, but... :shrug: doesn't cost much.

I think my main feedback would be to add more features to what can be detected. At the moment it may seem redundant to have older features, since software versions tend to be rather linear and this request would be a newer feature than e.g. ?format=raw. But having optional features that are not assumed to be present by the rest of the ecosystem lowers the barrier to entry for new gateway implementations.

Reply:
Note to self: i will move this to a separate IPIP, and we will continue there

Reply:
@lidel yea this came up already as something we'd like to have when were adding support for range requests to Lassie -- we need to be able to figure out if upstream providers support them (web3.storage currently does NOT).

Reply:
Not sure OPTIONS will help with the immediate problem above, but just saying this sort of "what features do you support?" is pretty useful and probably will remain so.

Reply:
@hannahhoward thank  you for flagging the need at hand. 
Extracted HTTP OPTIONS discussion to a separate IPIP https://github.com/ipfs/specs/pull/425 – see https://github.com/ipfs/specs/pull/425#discussion_r1254881047 for range request example.